src/ClientCredentialsHandler.cs

```diff
@@ -1,114 +1,94 @@
+        {
+            SetupClientCredentialsHandler(configuration, httpClientFactory);
+        }
+
+        private void SetupClientCredentialsHandler(IClientCredentialsOptions configuration, IHttpClientFactory httpClientFactory = null)
+        {
             if (configuration == null)
             {
                 var nullConfigException = new ArgumentNullException(nameof(configuration));
-                _logger?.LogError(nullConfigException.Message);
                 throw nullConfigException;
             }
 
-            var (isValid, whyInvalid) = configuration.IsValid();
-            if (!isValid)
-            {
-                var invalidConfigException = new ArgumentException(string.Join("\n", whyInvalid));
-                _logger?.LogError(invalidConfigException.Message);
-                throw invalidConfigException;
-            }
+            configuration.IsValid();
             Configuration = configuration;
-            
+
             if (httpClientFactory != null)
             {
-                _httpClient = httpClientFactory.CreateClient();
-            } else
-            {
-                ConfigureHttpClient();
+                HttpClient = httpClientFactory.CreateClient();
             }
-        }
-        
-        private void ConfigureHttpClient()
-        {
-            if (_httpClient == null)
+            else
             {
-                _httpClient = new HttpClient(new HttpClientHandler()
+                HttpClient = new HttpClient(new HttpClientHandler()
                 {
-                    AllowAutoRedirect = true,
+                    AllowAutoRedirect = false,
                     UseCookies = false,
                 });
             }
         }
 
-        public async Task<string> GetAccessToken()
+        public async Task<(string accessToken, string refreshToken)> GetAccessTokenAsync(CancellationToken cancellationToken = default)
         {
-            // Generate auth header
-            var authHeader = $"Bearer {JwtUtil.CreateClientCredentialsAuthorizationJwt(Configuration)}";
-
-            // Get token from oauth
-            var jwt = await RequestAccessToken(authHeader);
-
-            return jwt;
+            return await RequestAccessTokenAsync(cancellationToken);
         }
 
-        public async Task<JsonWebToken> GetAccessTokenAsJWT()
+        public async Task<(string accessToken, string refreshToken)> RefreshAccessTokenAsync(string refreshToken, CancellationToken cancellationToken = default)
         {
-            string access_token = await GetAccessToken();
-            var jwt = JwtUtil.ReadJWT(access_token);
-            return jwt;
+            return await RequestAccessTokenAsync(cancellationToken);
         }
 
-        private async Task<string> RequestAccessToken(string authHeader)
+        private async Task<(string accessToken, string refreshToken)> RequestAccessTokenAsync(CancellationToken cancellationToken = default)
```

Comment:
It would be nice if we use named tuple here so we could distinguish access token and refresh token.
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-tuples#tuple-field-names